### DISCUSSION TITLE
Question about custom property in UserControl

### DISCUSSION THREAD

**Discussion by sht2017** [AvaloniaUI/Avalonia#9754] [Q&A] [answered]
Here is my code. I'm confused why it can't work.
```
    public class CapsuleControlViewModel : TemplatedControl
    {


        public static readonly StyledProperty<IImage> ImgSourceProperty =
            AvaloniaProperty.Register<Image, IImage>("Test");

        public IImage ImgSource
        {
            get { return GetValue(ImgSourceProperty); }
            set { SetValue(ImgSourceProperty, value); }
        }
    }
```
When I try to use the property ```Test``` in my window like below, the error will occur.
```<uc:CapsuleControl Width="668" Test="avares://Deck Toolkit/Assets/lib_capsule.png"/>``` (The uc is the namespace of usercontrols)
I've looked at the [official tutorial](https://docs.avaloniaui.net/docs/authoring-controls/defining-properties), but I'm still confused.
And it seems like I'm not able to use the ```nameof()``` method, when I tried that, error [CS0103](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-messages/cs0103) would happen.

**Comment by BAndysc** [comment, marked as answer]
1. In the `Register` method you should pass your control type as the first generic argument (`public static readonly StyledProperty<IImage> ImgSourceProperty = AvaloniaProperty.Register<CapsuleControlViewModel, IImage>("Test");`)
2. The name of the property and the name you provide as the first argument should match (you pass `Test` as the property name, but the property is named `ImgSource`.

(3) This is not related to your problem, but having a TemplatedControl which is named ViewModels sounds like some MVVM pattern violation, `TemplatedControl` is a view, a viewmodel should be another class. But again, that's just architecture, not the thing you are asking for.
(4) Btw. this is a `TemplatedControl`, not an `UserControl`

**Reply by sht2017** [reply]
Thanks, actually the major problem is the third one. I forgot I was in the ViewModel file. Once I switched it to the class file, it worked. And you also noticed the class extended by ```CapsuleControlViewModel ``` is ```TemplatedControl``` and it is meant to be the ```ViewModelBase```. Sorry for my stupid, and for the last, thank you so much :)
